Iniciando tutorial sobre como criar um coletor (#32)
dadosjusbr/coletores: TUTORIAL.md

# Criando um um coletor de remunerações do sistema justiça

> Um esclarecimento importante é que o coletor de remunerações não apenas coleta. Também é responsabilidade do coletor traduzir o conteúdo dos arquivos baixados em um [resultado de coleta](https://github.com/dadosjusbr/storage/blob/master/agency.go#L27). Isso não quer dizer que a pessoa que decidir contribuir precisa criar um único [PR](https://help.github.com/pt/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request) com a coleta e tradução, pelo contrário. Nós recomendamos que a funcionalidade seja enviada para revisão em, pelo menos, duas etapas.

Reviewer: Entendi o quis dizer, mas acho que a redação ficou um pouco confusa. Especialmente como introdução do tutorial.

```suggestion
Na nomenclatura do DadosJusBR, um coletor (_crawler_) de remunerações é responsável por duas tarefas: baixar os dados do site oficial do órgão e convertê-los para o formato padronizado de [resultado de coleta](https://github.com/dadosjusbr/storage/blob/master/agency.go#L27) (_crawling result_). Para facilitar o processo de revisão, aconselhamos que separe o coletor em dois [PRs](https://help.github.com/pt/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request), sendo o segundo aberto depois do primeiro ter sido aprovado. Assim, acreditamos que o processo de revisão fica mais simples e rápido.
```

Introduzi nomes em inglês em parênteses porque acredito que facilitaria a tradução e referência quando abri o arquivo linkado.

> Outro esclarecimento importante é que o DadosJusBR se refere a órgãos do *sistema de justiça* e não apenas do judiciário. Também estamos interessados em órgãos regionais e federais. Isso quer dizer que é do interesse e escopo do projeto libertar remunerações de tribunais (por exemplo, TRF1, TRT13, TREPB e TJPB), Ministérios Públicos (por exemplo, MPF e MPPB), Defensorias e Procuradorias.

Reviewer: ```suggestion
Outro esclarecimento importante é que buscamos todos os órgãos do *sistema de justiça* e não apenas do Judiciário. Também estamos interessados em órgãos regionais e federais. Isso quer dizer que é do interesse e escopo do projeto libertar remunerações de tribunais (por exemplo, TRF1, TRT13, TREPB e TJPB), Ministérios Públicos (por exemplo, MPF e MPPB), Defensorias e Procuradorias.
```

Reviewer 2: Concordo com a sugestão de Irio

Feitos os devidos esclarecimentos preliminares, mãos a obra!

## Design e API

Um coletor é um [processo](https://pt.wikipedia.org/wiki/Processo_(inform%C3%A1tica) com uma [API](https://pt.wikipedia.org/wiki/Interface_de_programa%C3%A7%C3%A3o_de_aplica%C3%A7%C3%B5es) específica:

A invocação do processo coletor recebe via linha de comando dois parâmetros:

- `--mes`: Mês ao qual se refere a coleta
- `--ano`: Ano ao qual se refere a coleta

E são esperadas as seguintes saídas:

- `stdout`: Em caso de sucesso, imprimir um [CrawlingResult](https://github.com/dadosjusbr/storage/blob/master/agency.go#L27) em formato JSON
- `stderr`: Em caso de sucesso, vazio. Em caso de erro, imprimir um relatório que facilite encontrar e procurar o erro
- `status`: Em caso de sucesso 0 (zero), ou diferente de 0 caso contrário

> Coletores podem precisar de mais configuração além do mês e do ano da coleta. Como visto [aqui](https://github.com/dadosjusbr/coletores/tree/master/mppb) e [aqui](https://github.com/dadosjusbr/coletores/tree/master/trepb), recomendamos a utilização de [variávias de ambiente](https://pt.wikipedia.org/wiki/Vari%C3%A1vel_de_ambiente) para esse fim.

Reviewer: ```suggestion
Coletores podem precisar de mais configuração além do mês e do ano da coleta. Como visto [aqui](https://github.com/dadosjusbr/coletores/tree/master/mppb) e [aqui](https://github.com/dadosjusbr/coletores/tree/master/trepb), recomendamos a utilização de [variáveis de ambiente](https://pt.wikipedia.org/wiki/Vari%C3%A1vel_de_ambiente) para esse fim.
```

> Para dar liberdade de desenvolvimento e facilitar o gerenciamento de dependências, coletores do DadosJusBR são executados em um [contêiner Docker](https://aws.amazon.com/pt/containers/?nc1=f_ccr). Dentre outras vantagens, isso permite que coletores sejam escritos diferentes linguagens. Sendo assim, você pode esperar que seu coletor será executado da seguinte forma:

Reviewer: ```suggestion
Para dar liberdade de desenvolvimento e facilitar o gerenciamento de dependências, coletores do DadosJusBR são executados em um [contêiner Docker](https://aws.amazon.com/pt/containers/?nc1=f_ccr). Dentre outras vantagens, isso permite que coletores sejam escritos em diferentes linguagens. Sendo assim, você pode esperar que seu coletor será executado da seguinte forma:
```

```sh
$ git clone https://github.com/dadosjusbr/coletores
$ cd coletores/nomeColetor

> $ docker build --build-arg GIT_COMMIT=$(git rev-list -1 HEAD) -t nomeColetor .

Reviewer: Me pergunto o porquê desse `GIT_COMMIT`. Faz sentido colocar uma nota sobre isso? Sem saber do motivo, acho que caberia um parágrafo explicando o que acontece depois do `docker run`, citando a motivação desse argumento.

Reviewer 2: So uma observação mesmo sobre o GIT_COMMIT, o meu coletor mesmo do MPPE não utilizei porque não vi uma necessidade de fato. É realmente importante o uso disso nos coletores?

Author: Sim, a utilização do `GIT_COMMIT` é muito importante e deveria ser feita pelo próprio ambiente. Mas, como não sabemos ainda se o momento da coleta será o mesmo momento do armazenamento dos dados, preferimos deixar isso mais próximo da coleta.

Atualizei o texto com a importância do identificador da versão e o que fazer com `GIT_COMMIT', bem como com os passos que vem depois da coleta.

Reviewer 2: hmmmmmmmmmm, entendi entendi e concordo, vou dar uma estudada melhor pra no próximo PR mandar com GIT_COMMIT

# Criando um um coletor de remunerações do sistema justiça

> Na nomenclatura do DadosJusBR, um coletor (_crawler_) de remunerações é responsável por duas tarefas: baixar os dados do site oficial do órgão e convertê-los para o formato padronizado de [resultado de coleta](https://github.com/dadosjusbr/storage/blob/master/agency.go#L27) (_crawling result_). Para facilitar o processo de revisão, aconselhamos que separe o coletor em dois [PRs](https://help.github.com/pt/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request), sendo o segundo aberto depois do primeiro ter sido aprovado. Assim, acreditamos que o processo de revisão fica mais simples e rápido.

Reviewer 3: O que deve ser submetido nos 2 PRs?
Me pareceu confuso porque é falado da construção de um coletor e depois que isso deve ser feito em 2 PRs, mas me falta a informação do que deve ser submetido.

Author: Especifiquei melhor o que deve ir em cada PR.

Feitos os devidos esclarecimentos preliminares, mãos a obra!

## Design e API

Um coletor é um [processo](https://pt.wikipedia.org/wiki/Processo_(inform%C3%A1tica) com uma [API](https://pt.wikipedia.org/wiki/Interface_de_programa%C3%A7%C3%A3o_de_aplica%C3%A7%C3%B5es) específica:

A invocação do processo coletor recebe via linha de comando dois parâmetros:

- `--mes`: Mês ao qual se refere a coleta
- `--ano`: Ano ao qual se refere a coleta

E são esperadas as seguintes saídas:

- `stdout`: Em caso de sucesso, imprimir um [CrawlingResult](https://github.com/dadosjusbr/storage/blob/master/agency.go#L27) em formato JSON
- `stderr`: Em caso de sucesso, vazio. Em caso de erro, imprimir um relatório que facilite encontrar e procurar o erro
- `status`: Em caso de sucesso 0 (zero), ou diferente de 0 caso contrário

> Coletores podem precisar de mais configuração além do mês e do ano da coleta. Como visto [aqui](https://github.com/dadosjusbr/coletores/tree/master/mppb) e [aqui](https://github.com/dadosjusbr/coletores/tree/master/trepb), recomendamos a utilização de [variávias de ambiente](https://pt.wikipedia.org/wiki/Vari%C3%A1vel_de_ambiente) para esse fim.

Reviewer 3: Será que facilita se exemplificar um coletor que precise de variável de ambiente?

Author: Tem dois coletores listados no texto.

Para dar liberdade de desenvolvimento e facilitar o gerenciamento de dependências, coletores do DadosJusBR são executados em um [contêiner Docker](https://aws.amazon.com/pt/containers/?nc1=f_ccr). Dentre outras vantagens, isso permite que coletores sejam escritos em diferentes linguagens. Sendo assim, você pode esperar que seu coletor será executado da seguinte forma:

```sh
$ git clone https://github.com/dadosjusbr/coletores
$ cd coletores/nomeColetor
$ docker build --build-arg GIT_COMMIT=$(git rev-list -1 HEAD) -t nomeColetor .

> $ docker run docker run --mount source=dadosjus,target=/dadojusbr_data/ --env-file=.env nomeColetor --mes=01 --ano=2020 > nomeColetor_2020_01.json

Reviewer 3: Eu acho que facilita a leitura se explicar os parâmentros que estão sendo passados.

```

> As pessoas desenvolvedoras de coletores devem sempre ter o cuidado de setar o campo [Crawler](https://github.com/dadosjusbr/storage/blob/master/agency.go#L31) (sim teremos um teste de integração que garanta isso). Esse campo possui atributos que identificam unicamente o coletor e a coleta. Essa informações são muito importantes pois além de facilitar a depuração/detecção de problemas, elas tornam as informações publicadas pelo DadosJusBR completamente auditáveis. A variável de ambiente `GIT_COMMIT` deve ser utilizada para setar a versão do coletor executado para extrair os dados ([Crawler.CrawlerID](https://github.com/dadosjusbr/storage/blob/master/agency.go#L22)).

Reviewer 2: **Essas** informações são muito importantes pois além de facilitar a depuração/detecção de problemas, elas tornam as informações publicadas pelo DadosJusBR completamente auditáveis.
